src/System.IO.FileSystem/src/System/IO/Directory.cs

```diff
@@ -264,31 +264,67 @@ public static void Move(string sourceDirName, string destDirName)
 
             string fullsourceDirName = Path.GetFullPath(sourceDirName);
             string sourcePath = PathInternal.EnsureTrailingSeparator(fullsourceDirName);
+            string sourceDirectoryName = Path.GetFileName(fullsourceDirName);
 
             string fulldestDirName = Path.GetFullPath(destDirName);
             string destPath = PathInternal.EnsureTrailingSeparator(fulldestDirName);
+            string destDirectoryName = Path.GetFileName(fulldestDirName);
 
-            StringComparison pathComparison = PathInternal.StringComparison;
+            bool sameDirectoryIgnoreCase =
+                string.Equals(sourcePath, destPath, StringComparison.OrdinalIgnoreCase);
+            bool sameDirectoryNameIgnoreCase
+                = !string.IsNullOrEmpty(sourceDirectoryName) && string.Equals(sourceDirectoryName, destDirectoryName, StringComparison.OrdinalIgnoreCase);
 
-            if (string.Equals(sourcePath, destPath, pathComparison))
+            // If the paths are the exact same, fail.
+            if (sameDirectoryIgnoreCase
+                && string.Equals(sourcePath, destPath, StringComparison.Ordinal))
                 throw new IOException(SR.IO_SourceDestMustBeDifferent);
 
             string sourceRoot = Path.GetPathRoot(sourcePath);
             string destinationRoot = Path.GetPathRoot(destPath);
-            if (!string.Equals(sourceRoot, destinationRoot, pathComparison))
+            // Don't check the paths if we already determined the paths are identical.
+            if (!sameDirectoryIgnoreCase &&
+                !string.Equals(sourceRoot, destinationRoot, StringComparison.OrdinalIgnoreCase))
                 throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
 
             // Windows will throw if the source file/directory doesn't exist, we preemptively check
             // to make sure our cross platform behavior matches NetFX behavior.
             if (!FileSystem.DirectoryExists(fullsourceDirName) && !FileSystem.FileExists(fullsourceDirName))
                 throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
 
-            if (FileSystem.DirectoryExists(fulldestDirName))
+            if (!sameDirectoryIgnoreCase // This check is to allowing renaming of directories
+                && FileSystem.DirectoryExists(fulldestDirName))
+                throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
+
+            if (!sameDirectoryIgnoreCase
```

Comment:
This is really complicated to have no comment. What exactly are you looking for? Why are you looking for the same directory with a different casing?